Tests/SwiftSyntaxMacroExpansionTest/LexicalContextTests.swift

```diff
@@ -0,0 +1,298 @@
+      guard let associatedValue = enumCase.parameterClause else {
+        return enumCase.name.text
+      }
+
+      let argumentNames = associatedValue.parameters.map { param in
+        guard let firstName = param.firstName else {
+          return "_:"
+        }
+
+        return firstName.text + ":"
+      }.joined()
+
+      return "\(enumCase.name.text)(\(argumentNames))"
+    }
+
+    // Accessors use their enclosing context, i.e., a subscript or pattern
+    // binding.
+    if self.is(AccessorDeclSyntax.self) {
+      guard let lexicalContext = context.lexicalContext.dropFirst().first else {
+        return nil
+      }
+
+      return lexicalContext.functionName(in: context)
+    }
+
+    // All declarations with identifiers.
+    if let identified = self.asProtocol(NamedDeclSyntax.self) {
+      return identified.name.trimmedDescription
+    }
+
+    // Extensions
+    if let extensionDecl = self.as(ExtensionDeclSyntax.self) {
+      // FIXME: It would be nice to be able to switch on type syntax...
+      let extendedType = extensionDecl.extendedType
+      if let simple = extendedType.as(IdentifierTypeSyntax.self) {
+        return simple.name.trimmedDescription
+      }
+
+      if let member = extendedType.as(MemberTypeSyntax.self) {
+        return member.name.trimmedDescription
+      }
+    }
+
+    // Pattern bindings.
+    if let patternBinding = self.as(PatternBindingSyntax.self),
+      let singleVarName = patternBinding.singleBindingName
+    {
+      return singleVarName
+    }
+
+    return nil
+  }
+}
+
+public struct FunctionMacro: ExpressionMacro {
+  public static func expansion<
+    Node: FreestandingMacroExpansionSyntax,
+    Context: MacroExpansionContext
+  >(
+    of node: Node,
+    in context: Context
+  ) -> ExprSyntax {
+    guard let lexicalContext = context.lexicalContext.first,
+      let name = lexicalContext.functionName(in: context)
+    else {
+      return #""<unknown>""#
+    }
+
+    return ExprSyntax("\(literal: name)")
+  }
+}
+
+public struct AllLexicalContextsMacro: DeclarationMacro {
+  public static func expansion(
+    of node: some FreestandingMacroExpansionSyntax,
+    in context: some MacroExpansionContext
+  ) throws -> [DeclSyntax] {
+    context.lexicalContext.compactMap { $0.as(DeclSyntax.self)?.trimmed }
+  }
+}
+
+final class LexicalContextTests: XCTestCase {
+  private let indentationWidth: Trivia = .spaces(2)
+
+  func testPoundFunction() {
+    assertMacroExpansion(
+      """
+      func f(a: Int, _: Double, c: Int) {
+        print(#function)
+      }
+
+      struct X {
+        var computed: String {
+          get {
+            #function
+          }
+        }
+
+        init(from: String) {
+          #function
+        }
+
+        subscript(a: Int) -> String {
+          #function
+        }
+
+        subscript(a a: Int) -> String {
+          #function
+        }
+      }
+
+      extension A {
+        static var staticProp: String = #function
+      }
+      """,
```

Comment:
I would prefer to split this up into multiple `assertMacroExpansion` calls, each with a single `#function`. Just makes it easier to debug it if it fails.